### ISSUE TITLE
Partial Fourier Error

### ISSUE THREAD

**Issue by neurolabusc** [ismrmrd/siemens_to_ismrmrd#24]
The partial fourier detection discussed in [issue 18](https://github.com/ismrmrd/siemens_to_ismrmrd/pull/18) is not correct. The current code

```
<xsl:choose>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 1">0.5</xsl:when>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 2">0.75</xsl:when>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 4">0.875</xsl:when>
	<xsl:otherwise>1.0</xsl:otherwise>
</xsl:choose>
```
   
is incorrect. The correct values are described in [GDCM](https://github.com/malaterre/GDCM/blob/master/Source/DataDictionary/CSAHeader.xml):

```
> the partial fourier information can take only 5 different values which are encoded as follows:
>
> 4/8 partial fourier = 0x1
> 5/8 = 0x2
> 6/8 = 0x4
> 7/8 = 0x8
> 8/8 = 0x10
```
Therefore, the code should be
```
<xsl:choose>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 1">0.5</xsl:when>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 2">0.625</xsl:when>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 4">0.75</xsl:when>
	<xsl:when test="siemens/MEAS/sKSpace/ucPhasePartialFourier = 8">0.875</xsl:when>
	<xsl:otherwise>1.0</xsl:otherwise>
</xsl:choose>
```
You can test this with a [sample dataset that includes 6/8 partial Fourier images](https://github.com/neurolabusc/dcm_qa/tree/master/In/TotalReadoutTime).

**Comment by neurolabusc** [comment]
@hansenms any reason this issue was closed without fixing
[IsmrmrdParameterMap_Siemens_EPI_FLASHREF.xsl](https://github.com/ismrmrd/siemens_to_ismrmrd/blob/c0f9afafd11101e7015711188a4bca896ab28c96/parameter_maps/IsmrmrdParameterMap_Siemens_EPI_FLASHREF.xsl#L41) and [IsmrmrdParameterMap_Siemens_T1Mapping_SASHA.xsl](https://github.com/ismrmrd/siemens_to_ismrmrd/blob/572560c9582139675a1267b5858ea0776b4085cc/parameter_maps/IsmrmrdParameterMap_Siemens_T1Mapping_SASHA.xsl#L42)?

**Comment by hansenms** [comment]
It has been open since 2017 with no repro and no suggested fix. If this is an issue we need a proper repro. Like an example dataset and a clear description of what your expected behavior is and what exactly you are seeing. Ideally there would be a PR to fix it. 

So I mostly closed it because I figured if it had been sitting since 2017 without a PR it was probably no longer relevant. I would propose a PR instead of an issue if you have a fix.

**Comment by neurolabusc** [comment]
Understood. Can I suggest you keep this issue closed but assign the label `wontfix`. I did provide a link to a repository that provides [sample images with different partial Fourier settings](https://github.com/neurolabusc/dcm_qa/tree/master/In/TotalReadoutTime/PF_PAT_POS_PFOV_PEres_INTERP_test2). However, it is completely understandable that addressing this is outside of the remit of the current developers. Both dcm2niix and GDCM provide correct solutions for any future users that need to accurately determine pF.

**Comment by hansenms** [comment]
I wouldn’t say it’s a won’t fix. Somebody just needs to take it on. There is a tendency that a lot of issues get opened with an expectation that some army of devs jump on and solve it. The reality is that you are probably in the best position to fix this and would have the test cases etc. Also in this case, the PR would probably be less text than this discussion. So if somebody would like a style sheet changes, let’s just get a PR.

**Comment by hansenms** [comment]
Actually, I am reopening this. It is not 100% clear these values are to be interpreted the same for all sequences and it may be intentional that they are the way that they are. However, we need to have a few more eyes on it. So I will put up a PR (#89) and have somebody else take a look at it.

In the future, it is a huge help if you just put up a PR for this sort of thing. Much quicker to move it through. Issues that require somebody else (who is not focused on it and has no tests cases) tend to just sit there.

**Comment by neurolabusc** [comment]
@hansenms I do believe that `ucPhasePartialFourier` is described similarly across all Siemens sequences - I have never heard of any errors from dcm2niix users. However, I would be happy to acquire some public validation datasets on our VE11C Siemens Prisma if someone can send me a sample of the `EPI_FLASHREF` or `T1Mapping_SASHA` DICOMs that I can Phoenix onto our system (send to the email listed in my avatar). I do host a number of public DICOM validation datasets [dcm_qa*](https://github.com/neurolabusc?tab=repositories&q=dcm_qa&type=&language=&sort=) that might be useful for testing the `siemens_to_ismrmrd` project. I have also sent an email to our Siemens Research Collaboration Manager to verify that `ucPhasePartialFourier` has a single definition across all sequences.

**Comment by hansenms** [comment]
Let's see if @kspaceKelvin can get us the information. It would not be the first time that a parameter is interpreted differently for different sequences. For phase resolution some sequences round up, others down, etc. It happens. That is why we have style sheets so that these things can be adjusted. The PR is up and let's see what Kelvin says and take it from there,

**Comment by neurolabusc** [comment]
@hansenms it would be great to have sample images that demonstrate the need for various style sheets. If we can isolate the situations, we can usually get support from the manufacturer engineers to develop a formula without a style sheet. This sounds a lot like the `Round_factor` that [GE uses](https://github.com/rordenlab/dcm2niix/tree/master/GE#total-readout-time) which was completely mysterious until the GE engineers described the formula they use. Even if style sheets are required, it would be nice to demonstrate these edge cases for the community, helping other tools like dicm2nii and dcm2niix.

**Comment by hansenms** [comment]
Yeah, images are not really what we are looking for here, since we care about the interpretation of the raw data, which may yield different images depending on the reconstruction.

In terms of getting to a formula, then I don’t think it’s possible or even desirable in general. There are some classes of sequences that tend to yield similar interpretation of the parameters so we can probably cover most sequences with a few style sheets but there will be variations. Moreover, custom research sequences may have a completely different interpretation which may be a total hack but having the stylesheets allows users to accommodate that and still get usable reconstructions. I don’t see any specific advantage in trying compile in a formulate. The stylesheets are a good trade off between between being prescriptive and flexibility.

That said, there are repetitive sections in stylesheets and some kind of include function would be nice so we don’t have to edit the same thing in many places.

And example datasets are of course great to illustrate different cases, so we could do more there. But critically this should not be images but raw datasets.

On GE they are in much much worse shape when it comes to getting close to a formula. And most GE raw datasets cannot be interpreted without actually “knowing” what the sequence did. There is nothing in the dataset to even apply a stylesheets to. That’s why the GE converter *also* requires a sequence specific compiled component to interpret the profile ordering. Rounding is the least of our worries there ;)

**Comment by kspaceKelvin** [comment]
Heh, my bad for not updating all of the stylesheets when I changed it in https://github.com/ismrmrd/siemens_to_ismrmrd/pull/18 .  This interpretation is correct for all sequences as of the current software version.
